src/tactic/lint/misc.lean

```diff
@@ -243,3 +244,20 @@ has been used. -/
   errors_found := "INCORRECT DEF/LEMMA" }
 
 attribute [nolint def_lemma] classical.dec classical.dec_pred classical.dec_rel classical.dec_eq
+
+/-- Checks whether the statement of a declaration is well-typed. -/
+meta def check_type (d : declaration) : tactic (option string) :=
+(type_check d.type >> return none) <|> return "The statement doesn't type-check"
+
+/-- A linter for missing checking whether statements of declarations are well-typed. -/
+@[linter]
+meta def linter.check_type : linter :=
+{ test := check_type,
+  auto_decls := ff,
```

Comment:
```suggestion
  auto_decls := tt,
```
Constructors are also auto decls and should be checked as well.

Reply:
If we do that, we cannot define structure on a new irreducible type by using the corresponding structure on the old type.
In particular: all of the following instances have equational lemmas that are ill-typed once `with_one` is irreducible.

```lean
@[to_additive] instance : monad with_one := option.monad
@[to_additive] instance : has_one (with_one α) := ⟨none⟩
@[to_additive] instance [has_mul α] : has_mul (with_one α) := ⟨option.lift_or_get (*)⟩
@[to_additive] instance : inhabited (with_one α) := ⟨1⟩
@[to_additive] instance [nonempty α] : nontrivial (with_one α) := option.nontrivial
@[to_additive] instance : has_coe_t α (with_one α) := ⟨some⟩
```

More failures are in this Gist: https://gist.github.com/fpvandoorn/015ce16dbd42f3ec33f26f44ef031c3a

Reply:
All of the errors here are equation lemmas and abstracted proofs terms.  (And `linarith.config.mk.inj*` which is weird, but also a meta definition.)

It's unfortunate that `auto_decls` covers so much.  My preference would be to set `auto_decls := tt` for the linter and exclude abstracted proofs and equation lemmas manually.  But I'm also happy if you merge the linter as it is now before it sits another week.

bors d+

Reply:
That is a good idea. I might do that in a follow-up PR.

bors merge